File: ts/project/.sdk/tm/kotlin/core/Utility.kt
package KOTLINPACKAGE.core

import KOTLINPACKAGE.utility.Register

// The transport fetch function: replaceable per-instance so features (retry,
// cache, netsim, proxy, test) can wrap it.
typealias FetcherFn = (Context, String, MutableMap<String, Any?>) -> Any?

/**
 * The utility object: every pipeline step as a replaceable function field,
 * mirroring the ts/go utility objects. Features and tests may replace
 * individual fields (e.g. wrap [fetcher]) on a per-instance basis.
 */
class Utility private constructor(register: Boolean) {

  lateinit var clean: (Context, Any?) -> Any?
  lateinit var done: (Context) -> Any?
  lateinit var makeError: (Context, RuntimeException?) -> Any?
  lateinit var featureAdd: (Context, Feature) -> Unit
  lateinit var featureHook: (Context, String) -> Unit
  lateinit var featureInit: (Context, Feature) -> Unit
  lateinit var fetcher: FetcherFn
  lateinit var makeFetchDef: (Context) -> MutableMap<String, Any?>
  lateinit var makeContext: (MutableMap<String, Any?>?, Context?) -> Context
  lateinit var makeOptions: (Context) -> MutableMap<String, Any?>
  lateinit var makeRequest: (Context) -> Response
  lateinit var makeResponse: (Context) -> Response
  lateinit var makeResult: (Context) -> Result
  lateinit var makePoint: (Context) -> Map<String, Any?>
  lateinit var makeSpec: (Context) -> Spec
  lateinit var makeUrl: (Context) -> String
  lateinit var param: (Context, Any?) -> Any?
  lateinit var prepareAuth: (Context) -> Spec
  lateinit var prepareBody: (Context) -> Any?
  lateinit var prepareHeaders: (Context) -> MutableMap<String, Any?>
  lateinit var prepareMethod: (Context) -> String
  lateinit var prepareParams: (Context) -> MutableMap<String, Any?>
  lateinit var preparePath: (Context) -> String
  lateinit var prepareQuery: (Context) -> MutableMap<String, Any?>
  lateinit var resultBasic: (Context) -> Result
  lateinit var resultBody: (Context) -> Result
  lateinit var resultHeaders: (Context) -> Result
  lateinit var transformRequest: (Context) -> Any?
  lateinit var transformResponse: (Context) -> Any?
  var custom: MutableMap<String, Any?> = linkedMapOf()

  constructor() : this(true)

  init {
    if (register) {
      Register.registerAll(this)
    }
  }

  /** A field-level copy sharing nothing mutable but the function refs. */
  fun copy(): Utility {
    val u = Utility(false)

    u.clean = this.clean
    u.done = this.done
    u.makeError = this.makeError
    u.featureAdd = this.featureAdd
    u.featureHook = this.featureHook
    u.featureInit = this.featureInit
    u.fetcher = this.fetcher
    u.makeFetchDef = this.makeFetchDef
    u.makeContext = this.makeContext
    u.makeOptions = this.makeOptions
    u.makeRequest = this.makeRequest
    u.makeResponse = this.makeResponse
    u.makeResult = this.makeResult
    u.makePoint = this.makePoint
    u.makeSpec = this.makeSpec
    u.makeUrl = this.makeUrl
    u.param = this.param
    u.prepareAuth = this.prepareAuth
    u.prepareBody = this.prepareBody
    u.prepareHeaders = this.prepareHeaders
    u.prepareMethod = this.prepareMethod
    u.prepareParams = this.prepareParams
    u.preparePath = this.preparePath
    u.prepareQuery = this.prepareQuery
    u.resultBasic = this.resultBasic
    u.resultBody = this.resultBody
    u.resultHeaders = this.resultHeaders
    u.transformRequest = this.transformRequest
    u.transformResponse = this.transformResponse

    u.custom = LinkedHashMap(this.custom)

    return u
  }
}

File: ts/project/.sdk/tm/csharp/core/Utility.cs
// ProjectName SDK - utility object: the pluggable set of pipeline functions.
// Every pipeline step is a delegate field so features (and custom utilities)
// can wrap or replace behaviour per client instance.

namespace ProjectNameSdk;

public class Utility
{
    public Func<Context, object?, object?> Clean = null!;
    public Func<Context, object?> Done = null!;
    public Func<Context, Exception?, object?> MakeError = null!;
    public Action<Context, Feature.BaseFeature> FeatureAdd = null!;
    public Action<Context, string> FeatureHook = null!;
    public Action<Context, Feature.BaseFeature> FeatureInit = null!;
    public FetcherFunc Fetcher = null!;
    public Func<Context, Dictionary<string, object?>> MakeFetchDef = null!;
    public Func<Dictionary<string, object?>?, Context?, Context> MakeContext = null!;
    public Func<Context, Dictionary<string, object?>> MakeOptions = null!;
    public Func<Context, Response> MakeRequest = null!;
    public Func<Context, Response> MakeResponse = null!;
    public Func<Context, Result> MakeResult = null!;
    public Func<Context, Dictionary<string, object?>?> MakePoint = null!;
    public Func<Context, Spec> MakeSpec = null!;
    public Func<Context, string> MakeUrl = null!;
    public Func<Context, object?, object?> Param = null!;
    public Func<Context, Spec> PrepareAuth = null!;
    public Func<Context, object?> PrepareBody = null!;
    public Func<Context, Dictionary<string, object?>> PrepareHeaders = null!;
    public Func<Context, string> PrepareMethod = null!;
    public Func<Context, Dictionary<string, object?>> PrepareParams = null!;
    public Func<Context, string> PreparePath = null!;
    public Func<Context, Dictionary<string, object?>> PrepareQuery = null!;
    public Func<Context, Result> ResultBasic = null!;
    public Func<Context, Result> ResultBody = null!;
    public Func<Context, Result> ResultHeaders = null!;
    public Func<Context, object?> TransformRequest = null!;
    public Func<Context, object?> TransformResponse = null!;
    public Dictionary<string, object?> Custom = new();

    public Utility()
    {
        Util.SdkUtility.RegisterAll(this);
    }

    private Utility(bool _noregister)
    {
    }

    // A shallow copy sharing the same delegates but with an independent
    // Custom map, so per-entity utility views can diverge safely.
    public static Utility Copy(Utility src)
    {
        var u = new Utility(true)
        {
            Clean = src.Clean,
            Done = src.Done,
            MakeError = src.MakeError,
            FeatureAdd = src.FeatureAdd,
            FeatureHook = src.FeatureHook,
            FeatureInit = src.FeatureInit,
            Fetcher = src.Fetcher,
            MakeFetchDef = src.MakeFetchDef,
            MakeContext = src.MakeContext,
            MakeOptions = src.MakeOptions,
            MakeRequest = src.MakeRequest,
            MakeResponse = src.MakeResponse,
            MakeResult = src.MakeResult,
            MakePoint = src.MakePoint,
            MakeSpec = src.MakeSpec,
            MakeUrl = src.MakeUrl,
            Param = src.Param,
            PrepareAuth = src.PrepareAuth,
            PrepareBody = src.PrepareBody,
            PrepareHeaders = src.PrepareHeaders,
            PrepareMethod = src.PrepareMethod,
            PrepareParams = src.PrepareParams,
            PreparePath = src.PreparePath,
            PrepareQuery = src.PrepareQuery,
            ResultBasic = src.ResultBasic,
            ResultBody = src.ResultBody,
            ResultHeaders = src.ResultHeaders,
            TransformRequest = src.TransformRequest,
            TransformResponse = src.TransformResponse,
            Custom = new Dictionary<string, object?>(src.Custom),
        };
        return u;
    }
}
